packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -2,104 +2,220 @@
+            // refund whatever wasn't deposited.
+            uint256 refund = amount.sub(amountDeposited);
+            (bool success, ) = msg.sender.call{value: refund}(''); //solhint-disable-line avoid-low-level-calls
+            require(success, 'Could not refund excess funds');
+        }
+    }
+
     /**
      * @notice Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
      * @dev Transfers as many funds escrowed against `channelId` as can be afforded for a specific destination. Assumes no repeated entries.
      * @param fromChannelId Unique identifier for state channel to transfer funds *from*.
-     * @param allocationBytes The abi.encode of AssetOutcome.Allocation
+     * @param outcomeBytes The encoded Outcome of this state channel
      * @param indices Array with each entry denoting the index of a destination to transfer funds to. An empty array indicates "all".
      */
     function transfer(
+        uint256 assetIndex, // TODO consider a uint48?
         bytes32 fromChannelId,
-        bytes calldata allocationBytes,
+        bytes memory outcomeBytes,
+        bytes32 stateHash,
+        address challengerAddress,
         uint256[] memory indices
-    ) external override {
+    ) public override returns (bytes memory) {
         // checks
         _requireIncreasingIndices(indices);
-        _requireCorrectAllocationHash(fromChannelId, allocationBytes);
+        _requireChannelFinalized(fromChannelId);
+        _requireMatchingFingerprint(
+            stateHash,
+            challengerAddress,
+            keccak256(outcomeBytes),
+            fromChannelId
+        );
+
+        Outcome.OutcomeItem[] memory outcome = abi.decode(outcomeBytes, (Outcome.OutcomeItem[]));
+        Outcome.AssetOutcome memory assetOutcome = abi.decode(
+            outcome[assetIndex].assetOutcomeBytes,
+            (Outcome.AssetOutcome)
+        );
+        require(
+            assetOutcome.assetOutcomeType == Outcome.AssetOutcomeType.Allocation,
+            '!allocation'
+        );
+        Outcome.AllocationItem[] memory allocation = abi.decode(
+            assetOutcome.allocationOrGuaranteeBytes,
+            (Outcome.AllocationItem[])
+        );
+        address asset = outcome[assetIndex].assetHolderAddress;
+
         // effects and interactions
-        _transfer(fromChannelId, allocationBytes, indices);
+        allocation = _transfer(asset, fromChannelId, allocation, indices); // update in place to newAllocation
+        outcome[assetIndex].assetOutcomeBytes = abi.encode(
+            Outcome.AssetOutcome(Outcome.AssetOutcomeType.Allocation, abi.encode(allocation))
+        );
+        bytes32 outcomeHash = keccak256(abi.encode(outcome));
+        _updateFingerprint(fromChannelId, stateHash, challengerAddress, outcomeHash);
+        emit FingerprintUpdated(fromChannelId, abi.encode(outcome));
+        return abi.encode(outcome);
```

Comment:
Do you know if there is a reason we return the new outcome in the `transfer` function but not the `claim` function?

Reply:
The reason is that we do not have a `claimAllAssets` function (yet). We need to return the outcome here to allow `transferAllAssets` to use the updated outcome from a first `transfer` to be passed in as an argument for the next `transfer` when looping over assets.

We probably want a `claimAllAssets` function. 

We will also want a more efficient way of paying out all assets that does not involve writing to storage so much.